File: fractional_knapsack.py
def fractionalKnapsack(items, W):
    items.sort(key=lambda x: x[0]/x[1], reverse=True)
    res = 0
    for item in items:
        if W-item[1] >= 0:
            W -= item[1]
            res += item[0]
        else:
            res += (item[0]/item[1])*W
    return res


profit = [60, 100, 120]
wt = [10, 20, 30]
W = 50
items = [(profit[i], wt[i]) for i in range(len(wt))]
print(fractionalKnapsack(items, W))

# Time Complexity: O(nlogn) -> Sorting the arr O(nlogn) + O(n) for interating in it
# Space Complexity: O(1)

File: FractionalKnapsack.java
import java.util.Arrays;
import java.util.Comparator;

public class FractionalKnapsack {
    public static double fractionalKnapsack(int W, int arr[][]) {
        double res = 0;
        Arrays.sort(arr, new Comparator<int[]>() {
            public int compare(int[] a, int[] b) {
                return b[0] / b[1] - a[0] / a[1];
            }
        });
        for (int item[] : arr) {
            if (W - item[1] < 0) {
                res += (item[0] * ((double) W / item[1]));
                break;
            }
            W -= item[1];
            res += item[0];
        }
        return res;
    }

    public static void main(String[] args) {
        int W = 50;
        int arr[][] = { { 60, 10 }, { 100, 20 }, { 120, 30 } };
        double maxVal = fractionalKnapsack(W, arr);
        System.out.println("Maximum value we can obtain = " + maxVal);
    }
}
